tests/test_eic.cc

```diff
@@ -0,0 +1,269 @@
+  double dp_p0 = 0.004; //0.0004
+  double n_ptcl = 5E8;
+  double sigma_s_ion = 2e-2;
+  Beam ion_beam(n_charge, n_mass, kinetic_energy, emit_nx0, emit_ny0, dp_p0, sigma_s_ion, n_ptcl);
+
+  //define the ring (not really used for friction force calculation)
+  string filename = CMAKE_SOURCE_DIR + std::string("/data/Booster.tfs");
+  
+  Lattice lattice(filename);
+  Ring ring(lattice, ion_beam);
+
+  //define the cooler
+  double cooler_length = 1;
+  double n_section = 1;
+  double magnetic_field = 5; //0.039
+  double beta_h = 10;
+  double beta_v = 17;
+  double dis_h = 0;
+  double dis_v = 0;
+  Cooler cooler(cooler_length,n_section,magnetic_field,beta_h,beta_v,dis_h, dis_v);
+
+  //define electron bunch
+  double sigma_x = 1.5e-2;
+  double sigma_y = 1.5e-2;
+  double sigma_s = 2e-2;
+  double n_particle = 1e7;
+  GaussianBunch gb(n_particle,sigma_x,sigma_y,sigma_s);
+            
+  double gamma_e = ion_beam.gamma();
+  double tmp_tr = 0.1; //0.05
+  double tmp_long = 0.1;
+  EBeam e_beam(gamma_e, tmp_tr, tmp_long, gb);
+    
+  //This choice tells ecool_rate_paras to select
+  // 4 transverse velocities spanning the whole space,
+  // from which we can whittle down to the one we are
+  // interested in. As yet, there is no way to select
+  // a single velocity in a straightforward way.
+  unsigned int n_tr = 4;
+  unsigned int n_long = 400;
+  EcoolRateParas ecool_rate_paras(n_tr, n_long);
+
+  double rate_x, rate_y, rate_s;
+            
+  ForceParas force_paras(ff);
+  force_paras.set_do_test(true);
+  ecooling_rate(ecool_rate_paras, force_paras, ion_beam, cooler, e_beam, ring, rate_x, rate_y, rate_s);
+
+  return;
+}
+
+//A method to sort rows of a matrix based on a specific column.
+// This modifies the original matrix
+void sortrows(std::vector<std::vector<double>>& matrix, int col) {    
+    std::sort(matrix.begin(),
+              matrix.end(),
+              [col](const std::vector<double>& lhs, const std::vector<double>& rhs) {
+                  return lhs[col] > rhs[col];
+              });
+}
+
+//A standard C++ CSV parser to handle the input files we'll need
+// for these comparisons
+vector< vector<double> > ReadCSV(string filename){
+    
+    std::ifstream infile(filename);
+
+    string line;
+    vector<double> vec;
+    vector< vector<double> > matrix;
+
+    while( std::getline(infile,line) ){
+        std::stringstream ss(line);
+        vector<double> row;
+        std::string data;
+        while (getline(ss,data,',') ){
+
+            //Handle lines that don't match the format
+            try{
+                row.push_back( stod(data) );
+            }
+            catch(const std::exception& e){
+             //It was a bad line, we'll toss the whole row   
+                continue;
+            }
+        }
+        //The variables are F_const, v_trans, v_long, electron density, force_trans, force_long
+        if (row.size()  == 6) matrix.push_back(row);
```

Comment:
Should this not blow up if there aren't exactly 6 entries?

Reply:
This is checking against a pre-made table with 6 columns. If the test doesn't generate 6 columns, it should blow up.

Reply:
The programming principle here is [Fail Fast](https://github.com/radiasoft/pykern/wiki/DesignHints#fail-fast). The condition in the code (`row.size() == 6`) is being tested, and the error case is being ignored until a later time, when it could be caught earlier. The advantage of catching earlier is that you are closer to the source of the problem. If the code doesn't blow up when generating the csv incorrectly, then blow up here. 

There's no cost to this from a performance point of view. The condition is reversed, and the code would look like:

```cpp
assert(row.size() == 6);
matrix.push_back(row);
```

This documents the contract between the parser and the generator for the end user at the earliest point it can document it.

Note: the earlier [`try-catch`](https://github.com/radiasoft/electroncooling/blob/d6fef3b4067cdacafc19c8e87dd9eead8cc0415a/tests/test_eic.cc#L122) should fail fast as well. Don't catch exceptions unless they are meaningfully processed by the program. In this case, ignoring the exception defers the error. If the exception is raised, it will show that it is processing the error. One meaningful way to catch this exception is to count the lines in the file, and then print an error when a line cannot be parsed. The error code should re-throw the exception in this case, because the code that uses ReadCSV assumes the result is valid.

Reply:
I'm remembering now that the file I was initially reading in was the output from a "jspec > file.txt" command, and as such had a couple of unrelated print statements at the head and couple at the end. This row.size() == 6 check was to exclude those rows to make sure I wasn't erroring out by sending characters into the GSL fit.  Since the files no longer have extra print statements in them, this suggested assert will work and I'll look for anywhere else I can fail fast.